File: crates/photo-tauri/src/lib/filter.ts
// 筛选/排序纯逻辑移植（GPUI 版 crates/photo-tool-app/src/state/filter.rs 的
// apply_filter_and_sort 过滤段 + domain.rs FilterCriteria::has_active_filter）。
// 前端持有全量 CaptureMeta，筛选/排序零 IPC（迁移计划 Q5）。
// 判定边界与 Rust 版逐一对应，测试见 filter.test.ts。
import type {
  CaptureMeta,
  FilterCriteria,
  ImageFormat,
  Rating,
  SortBy,
  SortDirection,
} from './bindings'

/** Rating → 数字（0–5），对齐 domain.rs 的 `as u8`（None=0 … Five=5） */
const RATING_ORDER: Rating[] = ['None', 'One', 'Two', 'Three', 'Four', 'Five']

function ratingValue(r: Rating): number {
  return RATING_ORDER.indexOf(r)
}

/** ImageFormat → 显示串，对齐 domain.rs 的 Display impl（primaryFormat 即其输出） */
export function formatToString(fmt: ImageFormat): string {
  switch (fmt) {
    case 'Jpeg':
      return 'JPEG'
    case 'Png':
      return 'PNG'
    case 'Tiff':
      return 'TIFF'
    case 'Heif':
      return 'HEIF'
    case 'WebP':
      return 'WebP'
    case 'Bmp':
      return 'BMP'
    case 'Gif':
      return 'GIF'
    case 'Other':
      return 'OTHER'
    default:
      return fmt.Raw // { Raw: string }：RAW 扩展名原样输出
  }
}

/**
 * 格式身份匹配：primaryFormat 与筛选格式小写归一后比较。
 * Rust 是精确字符串比较（meta.primary_format != fmt.to_string()），但两套
 * 数据大小写不一致——真实数据为 Display 大写（"JPEG"），mock 为小写
 * （"jpeg"）——小写比较保持「格式身份相等」这一判定边界，且不同格式的
 * 字符串不会因归一而互相碰撞。
 */
function matchesFormat(primaryFormat: string, fmt: ImageFormat): boolean {
  return primaryFormat.toLowerCase() === formatToString(fmt).toLowerCase()
}

/** 拍摄时间 → 日期串 YYYY-MM-DD；仅接受 filter.rs 的两种 NaiveDateTime 格式 */
const DATE_TAKEN_RE = /^(\d{4}-\d{2}-\d{2})[T ]\d{2}:\d{2}:\d{2}$/

function parseDateTaken(dateStr: string): string | null {
  const m = DATE_TAKEN_RE.exec(dateStr)
  return m ? m[1] : null
}

/**
 * 焦距字符串 → 毫米数值（如 "600mm" → 600、"840mm" → 840）。
 * 变焦镜头取首个数值（广角端，如 "70-200mm" → 70）——对"焦距≥X"的鸟类拍摄
 * 场景判界保守且可预期。解析失败返回 null（设了焦距区间时按"不匹配"排除）。
 */
export function parseFocalLengthMm(s: string): number | null {
  const m = /(\d+(?:\.\d+)?)/.exec(s)
  return m ? parseFloat(m[1]) : null
}

/**
 * 过滤：返回通过全部条件的下标数组（升序）。边界与 filter.rs 逐一对应：
 * - dateTaken 为 null 且设了日期范围 → 排除；dateTaken 存在但解析失败 → 保留
 *   （Rust 的 `if let Ok` 无 else 分支，解析失败直接落到下一条件）；
 * - minRating：无评分（None=0）不满足 ≥1；
 * - NotRecognized = 无识别记录（recognitionStatus 为 null）。
 */
export function filterCaptures(items: CaptureMeta[], criteria: FilterCriteria): number[] {
  const out: number[] = []
  for (let i = 0; i < items.length; i++) {
    const meta = items[i]
    // format_filter：格式精确匹配（大小写归一）
    if (criteria.formatFilter !== null && !matchesFormat(meta.primaryFormat, criteria.formatFilter)) {
      continue
    }
    // bird_names：鸟种多选，命中任一选中项即保留
    if (criteria.birdNames.length > 0) {
      if (meta.birdName === null || !criteria.birdNames.includes(meta.birdName)) continue
    }
    // date_from / date_to
    if (criteria.dateFrom !== null || criteria.dateTo !== null) {
      if (meta.dateTaken !== null) {
        const d = parseDateTaken(meta.dateTaken)
        if (d !== null) {
          if (criteria.dateFrom !== null && d < criteria.dateFrom) continue
          if (criteria.dateTo !== null && d > criteria.dateTo) continue
        }
        // 解析失败：Rust 行为 = 保留（无 else 分支）
      } else {
        continue // 无拍摄时间且设了日期范围 → 排除
      }
    }
    // unflagged_filter：只显示没有旗标的照片
    if (criteria.unflaggedFilter && meta.flag !== null) continue
    // min_rating：评分 ≥ N
    if (criteria.minRating !== null && ratingValue(meta.rating) < ratingValue(criteria.minRating)) {
      continue
    }
    // color_label：颜色标签精确匹配
    if (criteria.colorLabel !== null && meta.colorLabel !== criteria.colorLabel) continue
    // flag_filter：旗标精确匹配（Pick/Reject）
    if (criteria.flagFilter !== null && meta.flag !== criteria.flagFilter) continue
    // iso 区间（闭区间 [isoMin, isoMax]；无 ISO 数据且设了区间 → 排除）
    if (criteria.isoMin !== null || criteria.isoMax !== null) {
      if (meta.iso === null) continue
      if (criteria.isoMin !== null && meta.iso < criteria.isoMin) continue
      if (criteria.isoMax !== null && meta.iso > criteria.isoMax) continue
    }
    // 焦距区间（闭区间 [focalMin, focalMax]，mm 数值；无焦距或解析失败 → 排除）
    if (criteria.focalMin !== null || criteria.focalMax !== null) {
      if (meta.focalLength === null) continue
      const mm = parseFocalLengthMm(meta.focalLength)
      if (mm === null) continue // 解析失败 = 不匹配（有筛选时排除）
      if (criteria.focalMin !== null && mm < criteria.focalMin) continue
      if (criteria.focalMax !== null && mm > criteria.focalMax) continue
    }
    // 镜头多选（精确匹配 EXIF lens 串；空 = 不限）
    if (criteria.lensFilter.length > 0) {
      if (meta.lens === null || !criteria.lensFilter.includes(meta.lens)) continue
    }
    // 关键词筛选：包含任一选中关键词即中（空 = 不限）
    if (criteria.keywordFilter.length > 0) {
      if (!meta.keywords.some((k) => criteria.keywordFilter.includes(k))) continue
    }
    // recognition_filter
    switch (criteria.recognitionFilter) {
      case 'Confirmed':
        if (meta.recognitionStatus !== 'Confirmed') continue
        break
      case 'NeedsReview':
        if (meta.recognitionStatus !== 'NeedsReview') continue
        break
      case 'Unrecognized':
        if (meta.recognitionStatus !== 'Unrecognized') continue
        break
      case 'NotRecognized':
        if (meta.recognitionStatus !== null) continue
        break
      case 'All':
        break
    }
    out.push(i)
  }
  return out
}

/** 字符串比较（对齐 Rust String::cmp：码点序，非 locale） */
function cmpStr(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0
}

/**
 * 两拍摄的排序比较（对齐 filter.rs 的 sort_by 比较器）。
 * Modified 以 dateTaken 为 mtime 代理：浏览器端无 fs::metadata，拍摄时间为空
 * = None 排最前（模拟 Rust Option<SystemTime> 的 None < Some 语义）。
 * EyeSharpness 为 T0 新增（Rust 侧无比较器，纯前端实现）：None 语义对齐 Modified
 * （None 排最前），有值按数值升序；降序由 applyFilterAndSort 外层反转统一处理。
 */
export function compareCaptures(sortBy: SortBy, a: CaptureMeta, b: CaptureMeta): number {
  switch (sortBy) {
    case 'FileName':
      return cmpStr(a.baseName.toLowerCase(), b.baseName.toLowerCase())
    case 'DateTaken':
      return cmpStr(a.dateTaken ?? '', b.dateTaken ?? '') // 对齐 unwrap_or("")
    case 'FileSize':
      return (a.fileSize ?? 0) - (b.fileSize ?? 0) // 对齐 unwrap_or(0)
    case 'Rating':
      return ratingValue(a.rating) - ratingValue(b.rating)
    case 'EyeSharpness': {
      const sa = a.eyeSharpness
      const sb = b.eyeSharpness
      if (sa === null && sb === null) return 0
      if (sa === null) return -1 // None 排最前（对齐 Modified 分支的 Option 语义）
      if (sb === null) return 1
      return sa - sb // 数值升序；降序由外层 sortDirection 处理
    }
    case 'Modified': {
      const ta = a.dateTaken
      const tb = b.dateTaken
      if (ta === null && tb === null) return 0
      if (ta === null) return -1
      if (tb === null) return 1
      return cmpStr(ta, tb)
    }
  }
}

/**
 * 过滤 + 排序 → 下标数组（display_order，即 captures.items 下标）。
 * Array.prototype.sort 在现代引擎为稳定排序，与 Rust 稳定 sort_by 一致。
 */
export function applyFilterAndSort(
  items: CaptureMeta[],
  options: { criteria: FilterCriteria; sortBy: SortBy; sortDirection: SortDirection },
): number[] {
  const { criteria, sortBy, sortDirection } = options
  const indices = filterCaptures(items, criteria)
  indices.sort((a, b) => {
    const cmp = compareCaptures(sortBy, items[a], items[b])
    return sortDirection === 'Ascending' ? cmp : -cmp // 对齐 Descending => cmp.reverse()
  })
  return indices
}

/** 默认筛选条件（对齐 FilterCriteria::default()） */
export function defaultFilterCriteria(): FilterCriteria {
  return {
    formatFilter: null,
    birdNames: [],
    dateFrom: null,
    dateTo: null,
    minRating: null,
    colorLabel: null,
    flagFilter: null,
    unflaggedFilter: false,
    recognitionFilter: 'All',
    isoMin: null,
    isoMax: null,
    focalMin: null,
    focalMax: null,
    lensFilter: [],
    keywordFilter: [],
  }
}

/**
 * 是否有任一筛选条件生效（无筛选时操作集 = 全部文件，批量文件操作应拒绝执行）。
 * 采用 domain.rs FilterCriteria::has_active_filter（含 colorLabel）——
 * GPUI 的 RootView::has_active_filters 漏了 color_label 字段，这里按权威实现。
 */
export function hasActiveFilters(criteria: FilterCriteria): boolean {
  return (
    criteria.formatFilter !== null ||
    criteria.birdNames.length > 0 ||
    criteria.dateFrom !== null ||
    criteria.dateTo !== null ||
    criteria.minRating !== null ||
    criteria.colorLabel !== null ||
    criteria.flagFilter !== null ||
    criteria.unflaggedFilter ||
    criteria.recognitionFilter !== 'All' ||
    criteria.isoMin !== null ||
    criteria.isoMax !== null ||
    criteria.focalMin !== null ||
    criteria.focalMax !== null ||
    criteria.lensFilter.length > 0 ||
    criteria.keywordFilter.length > 0
  )
}

File: crates/photo-ui/src/model/filter.rs
//! 筛选纯逻辑（对应 §4.1）。

use chrono::NaiveDate;
use photo_domain::{
    CaptureMeta, ColorLabel, Flag, ImageFormat, Rating, RecognitionFilter, RecognitionStatus,
};
use serde::{Deserialize, Serialize};

/// UI 级筛选条件（包含 photo_domain::FilterCriteria 全部字段 + ISO/焦距/镜头/关键词筛选）
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct FilterCriteria {
    /// 格式精确匹配（大小写归一；RAW 为通配语义）
    pub format_filter: Option<ImageFormat>,
    /// 物种多选（命中任一即保留，空 = 不限）
    pub taxon_names: Vec<String>,
    /// 拍摄日期范围（闭区间，比较 YYYY-MM-DD）
    pub date_from: Option<NaiveDate>,
    pub date_to: Option<NaiveDate>,
    /// 最低评分（>= N 语义）
    pub min_rating: Option<Rating>,
    /// 颜色标签精确匹配
    pub color_label: Option<ColorLabel>,
    /// 旗标精确匹配（Pick/Reject）
    pub flag_filter: Option<Flag>,
    /// 只显示无旗标（与 flag_filter 互斥）
    pub unflagged_filter: bool,
    /// 只看**源文件为空**（0 字节）的照片：复制/传输中断的残骸，便于筛出来重拷或删除
    pub empty_source_only: bool,
    /// 识别状态过滤
    pub recognition_filter: RecognitionFilter,
    /// ISO 闭区间 [iso_min, iso_max]
    pub iso_min: Option<u32>,
    pub iso_max: Option<u32>,
    /// 焦距闭区间 [focal_min, focal_max]（mm）
    pub focal_min: Option<f64>,
    pub focal_max: Option<f64>,
    /// 镜头多选（精确匹配 EXIF lens 串，空 = 不限）
    pub lens_filter: Vec<String>,
    /// 关键词筛选（包含任一选中关键词即中，空 = 不限）
    pub keyword_filter: Vec<String>,
}

/// 从当前目录照片收集「镜头」筛选候选项（去重 + 排序；空白忽略）。
///
/// 候选只用于下拉展示；筛选语义仍由 matches_criteria 的 lens_filter 精确匹配决定
/// （候选里没有的旧筛选值不会被悄悄清掉，见 §13.3）。
pub fn lens_options(items: &[CaptureMeta]) -> Vec<String> {
    let mut out: Vec<String> = items
        .iter()
        .filter_map(|m| m.lens.as_deref())
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .map(str::to_string)
        .collect();
    out.sort();
    out.dedup();
    out
}

/// 从当前目录照片收集「物种」筛选候选项：顶层展示名 + 每个主体的展示名。
///
/// 为什么必须带主体名：筛选语义是「任一主体命中即保留」（见 matches_criteria），
/// 只列顶层名会让多主体照片的次要物种根本选不到。占位名「未识别」不作为候选。
pub fn taxon_options(items: &[CaptureMeta]) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    for meta in items {
        if let Some(name) = meta
            .taxon_name
            .as_deref()
            .map(str::trim)
            .filter(|n| !n.is_empty())
        {
            out.push(name.to_string());
        }
        for subject in &meta.subjects {
            let name = subject.display_name.trim();
            if !name.is_empty() && name != "<未识别>" {
                out.push(name.to_string());
            }
        }
    }
    out.sort();
    out.dedup();
    out
}

impl From<FilterCriteria> for photo_domain::FilterCriteria {
    fn from(c: FilterCriteria) -> Self {
        photo_domain::FilterCriteria {
            format_filter: c.format_filter,
            taxon_names: c.taxon_names,
            date_from: c.date_from,
            date_to: c.date_to,
            min_rating: c.min_rating,
            color_label: c.color_label,
            flag_filter: c.flag_filter,
            unflagged_filter: c.unflagged_filter,
            recognition_filter: c.recognition_filter,
        }
    }
}

/// Rating 转数值 (0..=5)，None=0, One=1, ..., Five=5
pub fn rating_value(r: Rating) -> u8 {
    match r {
        Rating::None => 0,
        Rating::One => 1,
        Rating::Two => 2,
        Rating::Three => 3,
        Rating::Four => 4,
        Rating::Five => 5,
    }
}

/// 非 RAW 格式的标准显示名集合
const NON_RAW_FORMAT_NAMES: &[&str] =
    &["jpeg", "png", "tiff", "heif", "webp", "bmp", "gif", "other"];

/// 格式匹配：RAW 是通配语义（若扩展名为 "RAW" 匹配任意非标准格式；若是具体扩展名如 "NEF" 精确匹配）
pub fn matches_format(primary_format: &str, fmt: &ImageFormat) -> bool {
    match fmt {
        ImageFormat::Raw(ext) => {
            let want = ext.to_ascii_uppercase();
            if want == "RAW" {
                !NON_RAW_FORMAT_NAMES
                    .iter()
                    .any(|&f| f.eq_ignore_ascii_case(primary_format))
            } else {
                primary_format.eq_ignore_ascii_case(&want)
            }
        }
        _ => primary_format.eq_ignore_ascii_case(&fmt.to_string()),
    }
}

/// 焦距字符串解析：提取第一个数值（如 "70-200mm" -> 70.0）
pub fn parse_focal_length_mm(s: &str) -> Option<f64> {
    let mut num_str = String::new();
    let mut found_num = false;
    let mut has_dot = false;

    for c in s.chars() {
        if c.is_ascii_digit() {
            num_str.push(c);
            found_num = true;
        } else if c == '.' && found_num && !has_dot {
            num_str.push(c);
            has_dot = true;
        } else if found_num {
            break;
        }
    }

    if found_num {
        num_str.parse::<f64>().ok()
    } else {
        None
    }
}

/// 拍摄时间解析为 NaiveDate
fn parse_date_taken(date_str: &str) -> Option<NaiveDate> {
    let s = date_str.trim();
    if s.len() < 10 {
        return None;
    }
    let b = s.as_bytes();
    if !b[0..4].iter().all(u8::is_ascii_digit)
        || (b[4] != b'-' && b[4] != b':')
        || !b[5..7].iter().all(u8::is_ascii_digit)
        || (b[7] != b'-' && b[7] != b':')
        || !b[8..10].iter().all(u8::is_ascii_digit)
    {
        return None;
    }
    let y: i32 = s[0..4].parse().ok()?;
    let m: u32 = s[5..7].parse().ok()?;
    let d: u32 = s[8..10].parse().ok()?;
    NaiveDate::from_ymd_opt(y, m, d)
}

/// 过滤照片：返回通过全部条件的下标数组（升序）
pub fn filter_captures(items: &[CaptureMeta], criteria: &FilterCriteria) -> Vec<usize> {
    let mut out = Vec::new();

    for (i, meta) in items.iter().enumerate() {
        // format_filter
        if let Some(fmt) = &criteria.format_filter {
            if !matches_format(&meta.primary_format, fmt) {
                continue;
            }
        }

        // taxon_names: 命中任一选中项即保留（多主体：任一主体的展示名命中即可）
        if !criteria.taxon_names.is_empty() {
            let matched = meta.subjects.iter().any(|s| {
                criteria.taxon_names.iter().any(|b| b == &s.display_name)
            }) || meta.taxon_name.as_ref().is_some_and(|name| {
                criteria.taxon_names.iter().any(|b| b == name)
            });
            if !matched {
                continue;
            }
        }

        // date_from / date_to
        if criteria.date_from.is_some() || criteria.date_to.is_some() {
            if let Some(date_taken) = &meta.date_taken {
                if let Some(d) = parse_date_taken(date_taken) {
                    if let Some(from) = criteria.date_from {
                        if d < from {
                            continue;
                        }
                    }
                    if let Some(to) = criteria.date_to {
                        if d > to {
                            continue;
                        }
                    }
                }
                // 解析失败：保留（对齐文档 §4.1）
            } else {
                // 无拍摄时间且设了日期范围 -> 排除
                continue;
            }
        }

        // unflagged_filter: 只显示无旗标
        if criteria.unflagged_filter && meta.flag.is_some() {
            continue;
        }

        // empty_source_only: 只看 0 字节的残骸
        if criteria.empty_source_only && !meta.is_empty_source() {
            continue;
        }

        // min_rating: 评分 >= N
        if let Some(min_r) = criteria.min_rating {
            if rating_value(meta.rating) < rating_value(min_r) {
                continue;
            }
        }

        // color_label
        if let Some(target_label) = criteria.color_label {
            if meta.color_label != target_label {
                continue;
            }
        }

        // flag_filter
        if criteria.flag_filter.is_some() && meta.flag != criteria.flag_filter {
            continue;
        }

        // iso 区间 [iso_min, iso_max]
        if criteria.iso_min.is_some() || criteria.iso_max.is_some() {
            match meta.iso {
                Some(iso) => {
                    if let Some(min_iso) = criteria.iso_min {
                        if iso < min_iso {
                            continue;
                        }
                    }
                    if let Some(max_iso) = criteria.iso_max {
                        if iso > max_iso {
                            continue;
                        }
                    }
                }
                None => continue,
            }
        }

        // 焦距区间 [focal_min, focal_max]
        if criteria.focal_min.is_some() || criteria.focal_max.is_some() {
            match &meta.focal_length {
                Some(fl) => match parse_focal_length_mm(fl) {
                    Some(mm) => {
                        if let Some(min_fl) = criteria.focal_min {
                            if mm < min_fl {
                                continue;
                            }
                        }
                        if let Some(max_fl) = criteria.focal_max {
                            if mm > max_fl {
                                continue;
                            }
                        }
                    }
                    None => continue,
                },
                None => continue,
            }
        }

        // lens_filter: 精确匹配
        if !criteria.lens_filter.is_empty() {
            let matched = match &meta.lens {
                Some(lens) => criteria.lens_filter.iter().any(|l| l == lens),
                None => false,
            };
            if !matched {
                continue;
            }
        }

        // keyword_filter: 包含任一关键词
        if !criteria.keyword_filter.is_empty() {
            let matched = meta
                .keywords
                .iter()
                .any(|k| criteria.keyword_filter.iter().any(|target| target == k));
            if !matched {
                continue;
            }
        }

        // recognition_filter
        match criteria.recognition_filter {
            RecognitionFilter::Confirmed => {
                if meta.recognition_status != Some(RecognitionStatus::Confirmed) {
                    continue;
                }
            }
            RecognitionFilter::NeedsReview => {
                if meta.recognition_status != Some(RecognitionStatus::NeedsReview) {
                    continue;
                }
            }
            RecognitionFilter::Unrecognized => {
                if meta.recognition_status != Some(RecognitionStatus::Unrecognized) {
                    continue;
                }
            }
            RecognitionFilter::NotRecognized => {
                if meta.recognition_status.is_some() {
                    continue;
                }
            }
            RecognitionFilter::All => {}
        }

        out.push(i);
    }

    out
}

/// 默认筛选条件
pub fn default_filter_criteria() -> FilterCriteria {
    FilterCriteria::default()
}

/// 是否有任一筛选生效（作为批量操作的安全边界）
pub fn has_active_filters(criteria: &FilterCriteria) -> bool {
    criteria.format_filter.is_some()
        || !criteria.taxon_names.is_empty()
        || criteria.date_from.is_some()
        || criteria.date_to.is_some()
        || criteria.min_rating.is_some()
        || criteria.color_label.is_some()
        || criteria.flag_filter.is_some()
        || criteria.unflagged_filter
        || criteria.empty_source_only
        || criteria.recognition_filter != RecognitionFilter::All
        || criteria.iso_min.is_some()
        || criteria.iso_max.is_some()
        || criteria.focal_min.is_some()
        || criteria.focal_max.is_some()
        || !criteria.lens_filter.is_empty()
        || !criteria.keyword_filter.is_empty()
}
